netket/hilbert/index/constrained.py

```diff
@@ -36,79 +54,212 @@ def compute_constrained_to_bare_conversion_table(
 
     @property
-    def _bare_numbers(self) -> np.ndarray:
-        """
-        Returns the conversion table between indices in the constrained space and
-        the corresponding unconstrained space.
-        """
-        if self.__bare_numbers is None:
-            self.__bare_numbers = compute_constrained_to_bare_conversion_table(
-                self._unconstrained_index, self._constraint_fn
-            )
+    def size(self):
+        return len(self.shape)
 
-        return self.__bare_numbers
+    def states_to_numbers(self, states: Array) -> Array:
+        self._setup()
+        return self._lookup_table.states_to_numbers(states)
 
-    def states_to_numbers(self, states):
-        out = self._unconstrained_index.states_to_numbers(states)
+    def numbers_to_states(self, numbers: Array):
+        self._setup()
+        return self._lookup_table.numbers_to_states(numbers)
 
-        out = np.searchsorted(self._bare_numbers, out)
+    def all_states(self):
+        self._setup()
+        return self._lookup_table.all_states()
 
-        if np.max(out, initial=0) >= self.n_states:
-            raise RuntimeError(
-                "The required state does not satisfy " "the given constraints."
-            )
+    def _compute_all_states(self):
+        return _sum_constrained_fock_all_states(
+            self.n_particles, self.shape, self.index_dtype
+        )
 
-        return out
+    def _setup(self):
+        if self._lookup_table is None:
+            lookup_table = LookupTableHilbertIndex(self._compute_all_states())
+            object.__setattr__(self, "_lookup_table", lookup_table)
 
-    def numbers_to_states(self, numbers):
-        if numbers.ndim != 1:
-            raise RuntimeError("Invalid input shape, expecting a 1d array.")
 
-        # convert to original space
-        numbers = self._bare_numbers[numbers]
-        out = np.empty((numbers.shape[0], self.size))
-        for i in range(numbers.shape[0]):
-            out[i] = self._unconstrained_index.number_to_state(numbers[i])
-        return out
+class SumConstrainedHilbertIndexStaticRange(SumConstrainedHilbertIndexFock):
+    # shape is uniform, with same StaticRange on all sites
+    _range: StaticRange
 
-    def all_states(self):
-        return self.numbers_to_states(np.arange(self.n_states))
+    def __init__(self, range_, size, sum_value, index_dtype=jnp.int64):
+        # sum_value: e.g. total_sz
+        self.shape = (range_.length,) * size
+        # convert the constraint to a constraint on the range [0,1,...,n]
+        self.n_particles = round((sum_value - range_.start * size) / range_.step)
+        self.index_dtype = index_dtype
+        self._range = range_
+
+    def _compute_all_states(self):
+        all_states_fock = super()._compute_all_states()
+        return self._range.index_to_basis(all_states_fock).astype(self.index_dtype)
+
+
+def ConstrainedHilbertIndex(local_states, size, constraint_fun, index_dtype=jnp.int64):
```

Comment:
This function is very complicated, and depending on the inputs you get outputs of different types (which is not a good pattern IMO).

I would suggest directly constructing a `SumConstrainedHilbertIndexStaticRange` where necessary (Fock, Spin, Fermions...) and keeping around a more general `ConstrainedHilbertIndex` for those who need it.

The function used to construct a `SumConstrainedHilbertIndexStaticRange` can do the check on the dimensionality and eventually create a generic `ConstrainedHilbertIndex`.